final/discuss-recommend/literate-prog.tex

```diff
@@ -1,11 +1,36 @@
-\section{Literate Programming}
-\label{sec:literate-programming-discuss}
-
-Another improvement is to extend the product with support for literate
-programming, as discussed in \cref{sec:literate-programming}. This would allow 
-for rapid prototyping and documentation at the same time~\cite{schulte2012},
-allowing developers of new DSLs to document and explain their language in an
-interactive manner that directly allows experimentation.
+\subsection{Implementing an IPython kernel for Jupyter notebooks}
+\label{sec:discuss-literate-programming}
+
+The client had expressed his interest in literate programming functionality for
+Spoofax, as explained in \cref{sec:literate-programming}. Implementing an
+IPython kernel on top of the core shell module should not take a lot of effort,
+however due to the problems discussed earlier priority was given to a solid
+working REPL and an Eclipse plugin. Therefore our product does not deliver an
+implementation of literate programming.
+
+IPython kernels provide a solid framework for literate programming that has
+proven itself in many different languages%
+\footnote{https://github.com/ipython/ipython/wiki/IPython-kernels-for-other-languages}.
+An IPython kernel is basically a daemon with several network sockets
+representing the in and output streams of the client application. Several
+message types are sent over these sockets in JSON format, such as requests to
+evaluate a certain string or requests regarding the kernel
+state\footnote{https://jupyter-client.readthedocs.io/ en/latest/kernels.html}.
+
+Since the invoker class from the core module just accepts any string given to
+it and then resolves the command itself, implementing an IPython kernel should
+be pretty straightforward. Most of the required work would be in creating an
+adequate messaging framework capable of understanding all defined JSON
+messages. When the messaging framework is in place user input can be sent
+directly to the invoker, after which the results can be returned via network
+sockets by visiting the ``IResult'' interfaces, just as with any other client
+implementation.
+
+By implementing an IPython kernel all functionality offered by Jupyter
+notebooks as shown in \cref{fig:ipython} would be available to the
```

Comment:
Won't this give the false impression that images and plots are then also available?